scripts/staging/hmm/HMM Training.py

```diff
@@ -0,0 +1,135 @@
+        sentence.append(word2)
+        no_of_words = no_of_words - 1
+    
+    while no_of_words > 0:
+        if (word0, word1, word2) in transitions.keys():
+            word3 = sample_word(transitions[(word0, word1, word2)])
+        else:
+            if (word0, word1) in transitions.keys():
+                word3 = sample_word(transitions[(word0, word1)])
+            elif (word0, word2) in transitions.keys():
+                word3 = sample_word(transitions[(word0, word2)])
+            elif (word1, word2) in transitions.keys():
+                word3 = sample_word(transitions[(word1, word2)])
+            else:
+                if word2 in second_word.keys():
+                    word3 = sample_word(second_word[word2])
+                else:
+                    word3_keys = np.random.choice(list(transitions.values()), 1)
+                    word3 = sample_word(word3_keys)
+
+        sentence.append(word3)
+        word0 = word1
+        word1 = word2
+        word2 = word3
+        no_of_words = no_of_words - 1
+    print(' '.join(sentence))
+
+def train_markov_model(data):
+    for line in data:
+        line_length = len(line)
+        for i in range(line_length):
+            token = line[i]
+            if i == 0:
+                initial_word[token] = initial_word.get(token, 0) + 1
+            else:
+                prev_token = line[i-1]
+                if i == 1:
+                    add2dict(second_word, prev_token, token)
+                elif i == 2:
+                    second_token = line[i-2]
+                    add2dict(transitions, (second_token, prev_token), token)
+                    add2dict(second_word, prev_token, token)
+                    add2dict(second_word, second_token, prev_token)
+                else:
+                    third_token = line[i-3]
+                    second_token = line[i-2]
+                    add2dict(transitions, (second_token, prev_token), token)
+                    add2dict(transitions, (third_token, second_token, prev_token), token)
+                    add2dict(transitions, (third_token, prev_token), token)
+                    add2dict(transitions, (third_token, second_token), prev_token)
+                    add2dict(second_word, prev_token, token)
+                    add2dict(second_word, second_token, prev_token)
+                    add2dict(second_word, third_token, second_token)
+    
+    # Normalize the distributions
+    initial_word_total = sum(initial_word.values())
+    for key, value in initial_word.items():
+        initial_word[key] = value / initial_word_total
+        
+    for prev_word, next_word_list in second_word.items():
+        second_word[prev_word] = list2probabilitydict(next_word_list)
+        
+    for word_pair, next_word_list in transitions.items():
+        transitions[word_pair] = list2probabilitydict(next_word_list)
+    
+
+
+data = pd.read_csv('text_matrix.csv', dtype=type('string') ,header=None).values
```

Comment:
On reading the data matrix I am running into the following error. The error occurs due to one extra entry for a specific text in your csv file. This could be fixed by making some changes in you hmm.py. 
Error: 
File "HMM Training.py", line 128, in <module>
    data = pd.read_csv('text_matrix.csv', dtype=type('string') ,header=None).values
  File "C:\Program Files\Python\Python37\lib\site-packages\pandas\io\parsers.py", line 685, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "C:\Program Files\Python\Python37\lib\site-packages\pandas\io\parsers.py", line 463, in _read
    data = parser.read(nrows)
  File "C:\Program Files\Python\Python37\lib\site-packages\pandas\io\parsers.py", line 1154, in read
    ret = self._engine.read(nrows)
  File "C:\Program Files\Python\Python37\lib\site-packages\pandas\io\parsers.py", line 2059, in read
    data = self._reader.read(nrows)
  File "pandas/_libs/parsers.pyx", line 881, in pandas._libs.parsers.TextReader.read
  File "pandas/_libs/parsers.pyx", line 896, in pandas._libs.parsers.TextReader._read_low_memory
  File "pandas/_libs/parsers.pyx", line 950, in pandas._libs.parsers.TextReader._read_rows
  File "pandas/_libs/parsers.pyx", line 937, in pandas._libs.parsers.TextReader._tokenize_rows
  File "pandas/_libs/parsers.pyx", line 2132, in pandas._libs.parsers.raise_parser_error
pandas.errors.ParserError: Error tokenizing data. C error: Expected 178 fields in line 6591, saw 179